### PR TITLE
home search

### PR DESCRIPTION
**Pull request by codextech**

### CODE REVIEW

File: src/event/event.service.ts [hunk 1]
```diff
@@ -259,6 +261,33 @@ export class EventService {
...
+    await this.getTenantSpecificEventRepository();
+    const { page, limit } = pagination;
+    const { search } = query;
+    const eventQuery = this.eventRepository
+      .createQueryBuilder('event')
+      .leftJoinAndSelect('event.user', 'user')
+      .leftJoinAndSelect('event.categories', 'categories')
+      .leftJoinAndSelect('event.group', 'group')
+      .leftJoinAndSelect('event.attendees', 'attendees')
+      .where('event.status = :status', { status: EventStatus.Published });
```
Comment (stanislavkhatko) [inline src/event/event.service.ts:277]: Search should also prioritise `EventVisibility.Published` and if authorised search through `EventVisibility.Authorized` as well

File: src/event/event.service.ts [hunk 2]
```diff
@@ -211,11 +212,12 @@ export class EventService {
...
     if (search) {
       eventQuery.andWhere(
-        `(event.name LIKE :search OR 
-          event.description LIKE :search OR 
-          event.location LIKE :search OR 
-          CAST(event.lat AS TEXT) LIKE :search OR 
-          CAST(event.lon AS TEXT) LIKE :search)`,
+        `(event.name LIKE :search OR
+          event.slug LIKE :search OR 
+          event.description LIKE :search OR
```
Comment (stanislavkhatko) [inline src/event/event.service.ts:217]: due to `description` field volume and missing index on that this query gonna be extremely long

File: src/home/home.service.ts [hunk 3]
```diff
@@ -79,4 +81,14 @@ export class HomeService {
       interests: interests,
     };
   }
+
+  async globalSearch(pagination: PaginationDto, query: HomeQuery) {
+    const event = await this.eventService.searchAllEvents(pagination, query);
+    const group = await this.groupService.searchAllGroup(pagination, query);
```
Comment (stanislavkhatko) [inline src/home/home.service.ts:87]: Would be better to have them async with Promise.all() as they may take time.

File: src/event/event.service.ts [hunk 4]
```diff
@@ -211,11 +212,12 @@ export class EventService {
...
+          CAST(event.type AS TEXT) LIKE :search OR
+          CAST(event.status AS TEXT) LIKE :search OR
```
Comment (stanislavkhatko) [inline src/event/event.service.ts:219]: search by type and status as well as slug isn't needed. Only name and filtering by status with visibility

**Review by stanislavkhatko — CHANGES_REQUESTED** [review]

File: src/event/event.service.ts [hunk 5]
```diff
@@ -259,6 +257,35 @@ export class EventService {
...
+    const { search } = query;
+    const eventQuery = this.eventRepository
+      .createQueryBuilder('event')
+      .leftJoinAndSelect('event.user', 'user')
+      .leftJoinAndSelect('event.categories', 'categories')
+      .leftJoinAndSelect('event.group', 'group')
+      .leftJoinAndSelect('event.attendees', 'attendees')
+      .where('event.status = :status', { status: EventStatus.Published })
+      .andWhere('event.visibility = :visibility', {
+        visibility: EventVisibility.Public,
```
Comment (stanislavkhatko) [inline src/event/event.service.ts:275]: This is questionable filter. This means user will never find non-public group in search. I'd remove it for now.

File: src/event/event.service.ts [hunk 6]
```diff
@@ -259,6 +257,35 @@ export class EventService {
...
+      });
+
+    if (search) {
+      eventQuery.andWhere(
+        `(event.name LIKE :search OR
+          event.slug LIKE :search)`,
```
Comment (stanislavkhatko) [inline src/event/event.service.ts:281]: As you know 'slug' index is already present, though user isn't searching for a slug but by 'name' instead. 
This functionality should include creation of 'name' index on event/group

File: src/group/group.service.ts [hunk 7]
```diff
@@ -264,7 +265,10 @@ export class GroupService {
       .leftJoinAndSelect('group.groupMembers', 'groupMembers')
       .leftJoinAndSelect('groupMembers.user', 'user')
       .leftJoinAndSelect('groupMembers.groupRole', 'groupRole')
-      .where('group.status = :status', { status: GroupStatus.Published });
+      .where('group.status = :status', { status: GroupStatus.Published })
+      .andWhere('group.visibility = :visibility', {
```
Comment (stanislavkhatko) [inline src/group/group.service.ts:269]: Pls remove visibility filter for now. Let's see user feedback first

File: src/group/group.service.ts [hunk 8]
```diff
@@ -291,12 +295,37 @@ export class GroupService {
...
     if (search) {
       groupQuery.andWhere(
-        `(group.name LIKE :search OR 
-          group.description LIKE :search OR 
-          group.location LIKE :search OR 
-          group.type LIKE :search OR 
-          CAST(group.lat AS TEXT) LIKE :search OR 
-          CAST(group.lon AS TEXT) LIKE :search)`,
+        `(group.name LIKE :search OR
+          group.slug LIKE :search`,
```
Comment (stanislavkhatko) [inline src/group/group.service.ts:299]: no need for slug

Comment (stanislavkhatko) [inline src/group/group.service.ts:299]: no slug search

File: src/home/home.service.ts [hunk 9]
```diff
@@ -79,4 +81,16 @@ export class HomeService {
       interests: interests,
     };
   }
+
+  async globalSearch(pagination: PaginationDto, query: HomeQuery) {
+    const [event, group] = await Promise.all([
+      this.eventService.searchAllEvents(pagination, query),
```
Comment (stanislavkhatko) [inline src/home/home.service.ts:87]: Well, I figured out now that you are pulling all nested entities on a simple global search request. All needed on frontend is event/group name and it's slug for now.

**Review by stanislavkhatko — CHANGES_REQUESTED** [review]

File: src/home/home.service.ts [hunk 10]
```diff
@@ -79,4 +81,27 @@ export class HomeService {
...
+      this.groupService.searchAllGroup(pagination, query),
```
Comment (stanislavkhatko) [inline src/home/home.service.ts:88]: Looks good, but all nested fields are still pulled from db.
Make a separate methods `showGlobalSearchEvents` and `showGlobalSearchGroups` and `select` only 'name and slug' instead.

**Review by stanislavkhatko — CHANGES_REQUESTED** [review]
Should be separate methods for global search instead of reusing page search functions. 
Missing db indexes for 'name' on group and event

File: src/event/event.service.ts (hunk 2 above)
Comment (tompscanlan) [reply src/event/event.service.ts]: Any ideas here @codextech ?  Do we need description here?  Do we add an index on description?  Do we ignore it and create an issue to come back and resolve ?

File: src/event/event.service.ts (hunk 6 above)
Comment (codextech) [reply src/event/event.service.ts]: its done

**Review by stanislavkhatko — CHANGES_REQUESTED** [review]
Please add simple unit and e2e method to wrap it up. Looks good

File: src/test/mocks.ts [hunk 11]
```diff
@@ -0,0 +1,429 @@
+import { CategoryEntity } from '../category/infrastructure/persistence/relational/entities/categories.entity';
```
Comment (stanislavkhatko) [inline src/test/mocks.ts:1]: @codextech overwriting the whole mocks file is nasty. Hope there are only fixes.

**Review by stanislavkhatko — APPROVED** [review]
Mocks file was overwritten, but if everything works then never mind

**Pull request merged**